packages/eslint-plugin/src/lib/util.ts

```diff
@@ -0,0 +1,144 @@
+import { Rule } from 'eslint';
+import { ParserServices } from '@typescript-eslint/typescript-estree';
+
+const version = require('../../package.json').version;
+
+export type ObjectLike<T = any> = Record<string, T>;
+
+export function tslintRule(name: string) {
+  return `\`${name}\` from TSLint`;
+}
+
+export function metaDocsUrl(name: string) {
+  return `https://github.com/typescript-eslint/typescript-eslint/blob/${version}/packages/eslint-plugin/docs/rules/${name}.md`;
+}
+
+/**
+ * Check if the context file name is *.ts or *.tsx
+ * @param fileName The context file name
+ * @returns `true` if the file name ends in *.ts or *.tsx
+ * @private
+ */
+export function isTypescript(fileName: string) {
+  return /\.tsx?$/i.test(fileName || '');
+}
+
+/**
+ * Check if the context file name is *.d.ts or *.d.ts
+ * @param fileName The context file name
+ * @returns `true` if the file name ends in *.d.ts or *.d.ts
+ * @private
+ */
+export function isDefinitionFile(fileName: string) {
+  return /\.d\.tsx?$/i.test(fileName || '');
+}
+
+/**
+ * Check if the variable contains an object stricly rejecting arrays
+ * @param obj an object
+ * @returns `true` if obj is an object
+ */
+function isObjectNotArray<T extends object>(obj: T | any[]): obj is T {
+  return typeof obj === 'object' && !Array.isArray(obj);
+}
+
+/**
+ * Pure function - doesn't mutate either parameter!
+ * Merges two objects together deeply, overwriting the properties in first with the properties in second
+ * @param first The first object
+ * @param second The second object
+ * @returns a new object
+ */
+export function deepMerge<T extends ObjectLike = ObjectLike>(
+  first: ObjectLike = {},
+  second: ObjectLike = {}
+): T {
+  // get the unique set of keys across both objects
+  const keys = new Set(Object.keys(first).concat(Object.keys(second)));
+
+  return Array.from(keys).reduce(
```

Comment:
again can probably `.reduce<T>` here instead of the cast further down

Reply:
In this case you can't because `{}` is not assignable to `T` without a forced `as` cast.